File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.cpp
class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extra) {
    std::priority_queue<std::tuple<double, int, int>> pq;

        for(auto&v:classes){
            // int gc = gcd(v[0],v[1]);
            // v[0] = (v[0]/gcd(v[0],v[1]));
            // v[1] = (v[1]/gcd(v[0],v[1]));
            //cout<<v[0]<<" "<<v[1]<<"\n";
            pq.push({(1.0*(v[0]+1)/(v[1]+1))-(1.0*v[0]/v[1]),v[0],v[1]});
        }

        while(extra&&pq.size()){
            auto [rat,a,b] = pq.top();
            pq.pop();

            double r = 1.0*(a+1)/(b+1);
            pq.push({(1.0*(a+2)/(2+b))-r,a+1,b+1});
            extra--;
        }

        double avg = 0;
        while(pq.size()){
            auto [rat,a,b] = pq.top();
            pq.pop();
            avg+=(1.0*a/b);
        }


        return avg/classes.size();
    }
};

File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.py
from queue import PriorityQueue


class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extra: int) -> float:
        pq = PriorityQueue()

        for ele in classes:
            pq.put([-(1.0*(1+ele[0])/(1+ele[1]))+(1.0*ele[0]/ele[1]),ele[0],ele[1]])
        
        while extra > 0:
            extra  = extra - 1
            ele = pq.get()
            pq.put([-(1.0*(2+ele[1])/(2+ele[2]))+(1.0*(1+ele[1])/(1+ele[2])),1+ele[1],1+ele[2]])

        ans  = 0.0
        while not pq.empty():
            ele = pq.get()
            ans = ans + ele[1]/ele[2]
        
        return 1.0*ans/len(classes)

        